components/brave_wallet/browser/zcash/zcash_orchard_sync_state.cc

```diff
@@ -0,0 +1,147 @@
+  return storage_->GetSpendableNotes(std::move(account_id));
+}
+
+base::expected<std::vector<OrchardNoteSpend>, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::GetNullifiers(mojom::AccountIdPtr account_id) {
+  return storage_->GetNullifiers(std::move(account_id));
+}
+
+base::expected<std::optional<uint32_t>, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::GetLatestShardIndex(mojom::AccountIdPtr account_id) {
+  return storage_->GetLatestShardIndex(std::move(account_id));
+}
+
+base::expected<std::optional<uint32_t>, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::GetMaxCheckpointedHeight(mojom::AccountIdPtr account_id,
+                                                uint32_t chain_tip_height,
+                                                size_t min_confirmations) {
+  return storage_->GetMaxCheckpointedHeight(
+      account_id.Clone(), chain_tip_height, min_confirmations);
+}
+
+std::optional<ZCashOrchardStorage::Error> ZCashOrchardSyncState::UpdateNotes(
+    mojom::AccountIdPtr account_id,
+    OrchardBlockScanner::Result block_scanner_results,
+    const uint32_t latest_scanned_block,
+    const std::string& latest_scanned_block_hash) {
+  auto existing_notes = storage_->GetSpendableNotes(account_id.Clone());
+  if (!existing_notes.has_value()) {
+    return existing_notes.error();
+  }
+
+  std::vector<OrchardNote> notes_to_add =
+      block_scanner_results.discovered_notes;
+  base::Extend(existing_notes.value(), notes_to_add);
+
+  std::vector<OrchardNoteSpend> nf_to_add;
+
+  for (const auto& nf : block_scanner_results.found_spends) {
+    if (std::find_if(existing_notes.value().begin(),
+                     existing_notes.value().end(), [&nf](const auto& v) {
+                       return v.nullifier == nf.nullifier;
+                     }) != existing_notes.value().end()) {
+      nf_to_add.push_back(nf);
+    }
+  }
+
+  if (!GetOrCreateShardTreeManager(account_id.Clone())
+           ->InsertCommitments(std::move(block_scanner_results))) {
+    return ZCashOrchardStorage::Error{
+        ZCashOrchardStorage::ErrorCode::kInternalError,
+        "Failed to insert commitments"};
+  }
+
+  return storage_->UpdateNotes(std::move(account_id), notes_to_add,
+                               std::move(nf_to_add), latest_scanned_block,
+                               latest_scanned_block_hash);
+}
+
+base::expected<bool, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::ResetAccountSyncState(mojom::AccountIdPtr account_id) {
+  return storage_->ResetAccountSyncState(std::move(account_id));
+}
+
+base::expected<std::vector<OrchardInput>, ZCashOrchardStorage::Error>
+ZCashOrchardSyncState::CalculateWitnessForCheckpoint(
+    mojom::AccountIdPtr account_id,
+    std::vector<OrchardInput> notes,
+    uint32_t checkpoint_position) {
+  auto result = GetOrCreateShardTreeManager(account_id.Clone())
+                    ->CalculateWitness(notes, checkpoint_position);
+  if (!result.has_value()) {
```

Comment:
`GetOrCreateShardTreeManager` returns a raw pointer so how can we call `has_value` on it?

Reply:
result has base::expected<OrchardNoteWitness, std::string>  type

Reply:
but `GetOrCreateShardTreeManager` can return `nullptr`

Reply:
oic now that it's calling `CalculateWitness` which I missed, but there is no `nullptr` check here for `GetOrCreateShardTreeManager` which also means there is a missing test to cover that scenario

Reply:
GetOrCreateShardTreeManager lazily creates ShardTreeManager for the provided account, so it shouldn't be null. I'll add CHECK